### DISCUSSION TITLE
HLSLPP_SCALAR and struct alignment?

### DISCUSSION THREAD

**Discussion by Benualdo** [redorav/hlslpp#90] [Ideas]
Hi, recently I add a bug on ARM and wanted to switch to scalar implementation of hlslpp onto that platform but this introduced another issue: mesh loading was not working anymore!

Turns out I was serializing a struct like this:

```
struct Vertex
{
    core::float4    pos; 
    core::float4    nrm;
    core::float4    bin;
    core::float4    tan;
    core::float4    color;
    core::float2    uv[2];
    u16             indices[4];
    u8              weights[4];
};
```

And sizeof(Vertex) was 128 with SSE implementation but only 124 with scalar implementation because of SSE alignement.

While I understand that serializing vector types is discutable (e.g., uv[0] is 16 bytes instead of 8 here) do you think it might be a good idea to enforce struct alignment of hlslpp structs like float4 even when the scalar implementation is used for consistency?

e.g.:

```
struct **alignas(16)** hlslpp_nodiscard	float4
{ 
```

The value of 16 here is because of SSE and could vary but at least we could have a define like 
`#define HLSLPP_MIN_ALIGNMENT 16`
?

I solved my issue by using alignas on my own struct, but what do you think about the feature for hlslpp?

Thanks,
Benoît.

**Comment by redorav** [comment]
Yes, that is a good idea. It should be consistent between platforms. I'll take a look

@Benualdo Also, what but did you get on ARM?

**Reply by Benualdo** [reply]
I could not spend much time investigating yet, but it seems that quaternion::rotation_euler_zxy returns different result in my ARM64EC build, I end up having a rotation quaternion that is not normalized. I will keep you posted and eventually open an issue when I can confirm the issue is not on my side. Thanks.

**Reply by Benualdo** [reply]
I ran unit tests on my Surface Pro 11 and they fail, 
![image](https://github.com/user-attachments/assets/1861d5db-1e43-4a2e-becc-8092dc98cc3e)
witch is weird, because values seem correct 
![image](https://github.com/user-attachments/assets/231e2b8c-8ae1-4330-882f-ea817cf75585)

**Reply by redorav** [reply]
Hmm ok I'll take a look when I can. Fwiw lerp is an odd one, I still need to see what the best course of action is. Because in floating point the order of operations matters, doing it as (B - A) * a + A is not the same as A * (1.0 - a) + B * a which I suspect is what you are seeing there. This happens when you turn on the FMA instruction set too for example. The implementation follows the HLSL fxc more closely, I don't know if fused multiply adds on GPUs actually respect that though

**Reply by Benualdo** [reply]
Using a small epsilon like `float1 vlerp1 = lerp(vfoo1, vbar1, vbaz1); hlslpp_check(eq(vlerp1, lerpf(vfoo1.r, vbar1.r, vbaz1.r), 0.000001f));` the test now passes indeed.

I'm more concerned about quaternion::rotation_euler_zxy returning crazy values on ARM:

MSVC x64:
![image](https://github.com/user-attachments/assets/111fc162-3164-48ea-9798-445013b55086)

MSVC ARM64 (Surface Pro 11):
![image](https://github.com/user-attachments/assets/bfbe8435-074c-460a-8a48-7ab116d09198)
(both debug and release)

Because that's exactly the issue I have in my engine.

**Reply by redorav** [reply]
Can you open separate issues for all those issues you have and I can take a look? With the inputs, the functions, essentially the repro

**Reply by redorav** [reply]
Disregard my last comment, I have been able to repro in Android as well. I'm trying to see where it goes wrong

**Reply by redorav** [reply]
Fixed in 4dd067be50f6a5dcb82880488a5615777c2533ca

**Reply by Benualdo** [reply]
Awesome, thanks!

**Comment by redorav** [comment]
This is fixed in 92729f2bebf1e8556ba3c4fb78a7638e251da494 btw

**Comment by redorav** [comment]
I think there is a further possible discussion regarding the size and alignment of the types. There is an argument towards having types with the same alignment across platforms if e.g. one platform is used to store data and the other to consume it. To be honest, I don't recommend using these types to store data due to wasted space but that could come with the interop structs. It is a known issue and it needs a solution.

I am enforcing the alignment of the float4 for now and we'll see what scope or desire there it to change it for other platforms (I'm thinking of NEON here)

**Reply by Benualdo** [reply]
Yeah, it's a common pitfall, but I don't like how many engines end up with two math libraries: one SIMD and one scalar. Maybe I'll just create 'storage' types like StorageFloat4, StorageFloat3, etc., with no methods except casts to and from HLSLPP types or just use float[] maybe. And IMHO, enforcing 16 bytes alignment for SIMD types is totally acceptable even for NEON because even if not stricly necessary it's better for performance but I understand other people could think otherwise, hence the #define for minimum enforced alignment idea.
